proposals/22866-unified-queue-implementation.md

```diff
@@ -0,0 +1,80 @@
+- Similarly, software installation requests are sent to `fleetd` via a different "Orbit notification" also returned by the `GET /config` orbit endpoint, and multiple software installation requests may be sent at once and will then be processed in that order by `fleetd`: https://github.com/fleetdm/fleet/blob/b4a5a1fb49666dd3b10cfd11ccf26190ad9d2902/server/service/orbit.go#L271-L277. Those are stored in the `host_software_installs` table (with `status` set to `pending`) for software packages installs/uninstalls. For App store apps, installation is via MDM commands, but the install request is saved in `host_vpp_software_installs` with the `command_uuid` pointing to the MDM command.
+
+## New single table for all upcoming activities
+
+The preferred implementation option at the moment is the table that sits in front of all actions that holds the arguments for these and the host they are intended for. This table is the unified queue for a host, and once items are processed they are removed and the next item in the list is marked as ready to process.
+
+Notable challenges with that approach are:
+- The table needs to be able to hold all the different types of actions and their arguments. Some actions require significant associated data (e.g. software installs can have a pre-install query, a custom install script and a post-install script).
+- The logic today expects data to be immediately stored in the relevant tables even while it is pending. With that approach, we'd store it in the unified queue until it is ready to be processed, and only then we'd store it in the relevant tables (e.g. `nano_commands` or `host_script_results` etc.). If not properly addressed this may cause subtle issues in behavior, UI, summary counts, etc.
+- Ensuring everything that is mutable still works after mutation (I don't think a lot of things are mutable - MDM commands certainly are not - but e.g. software install, what if the installer's information is updated?).
+- Triggering execution of the next item in the queue - push or pull? In most cases, when we save the result of the previous item we could trigger (push) execution of the next one, but what about the very first item in the queue, or what if there are no more items at this moment?
+	* An additional challenge is that even when the next item is created in the "delivery" table, it is still pending for a bit (e.g. for a script execution, pending until the host requests its config, gets the script to execute, and starts executing), so we should probably not remove it from the unified queue until it has been processed, but we should mark it as "in processing" so that it is not seen as the next item to process...
+
```

Comment:
:+1: yeah, I do call out the setup experience impact here, there may be more though: https://github.com/fleetdm/fleet/pull/23501/files#diff-f2cc7af549e313ccbf7a6122cbcdaf299c43bf20e0eccc532c86134a3b5ba521R58

Reply:
There's maybe some subtle, out-of-band orchestration needed to install the fleetd app (as well as the fleetd config profile), any bootstrap package app, run the setup experience script, and send the release device command.

Reply:
Related comments:
- currently we retry sending MDM command 1 time if it failed
- For optimal performance, MDM should give the host the next command after it checks in with the results of a previous command. If we miss that window, than we need Apple to send a push notification to tell the host to check in and get its next command.

Reply:
> currently we retry sending MDM command 1 time if it failed
That's a good call-out, because it adds some complexity to the "on receive results, send the next thing to execute"... We can't just blindly send the next thing to run, we need to do it only when we receive results for the last attempt (unless the result is success, in which case first result triggers the next action).

> If we miss that window, than we need Apple to send a push notification to tell the host to check in and get its next command
:100: